src/compiler/crystal/tools/tracing.cr

```diff
@@ -0,0 +1,337 @@
+      getter min : T = T::MAX
+      getter max : T = T::MIN
+      @sum_square : T
+
+      def initialize
+        @sum_square = T.zero
+      end
+
+      def <<(value)
+        v = T.new(value)
+        @size += 1
+        @sum += v
+        @min = v if v < @min
+        @max = v if v > @max
+        @sum_square += v ** 2
+      end
+
+      def average
+        size > 0 ? sum / size : T.zero
+      end
+
+      def stddev
+        if size > 0
+          Math.sqrt((@sum_square / size) - (average ** 2))
+        else
+          T.zero
+        end
+      end
+
+      def humanize(io, value)
+        io << value.round(9)
+      end
+
+      def to_s(io : IO) : Nil
+        io << "[total="
+        humanize io, sum
+        io << " min="
+        humanize io, min
+        io << " max="
+        humanize io, max
+        io << " mean="
+        humanize io, average
+        io << " ±"
+        humanize io, stddev
+        io << ']'
+      end
+    end
+
+    class Durations < Values(Float64)
+      def humanize(io, value)
+        value = value.abs
+
+        if value >= 1
+          io << value.round(2)
+          io << 's'
+        elsif value > 0.001
+          io << (value * 1_000).to_i64
+          io << "ms"
+        elsif value > 0.000_001
+          io << (value * 1_000_000).to_i64
+          io << "us"
+        else
+          io << (value * 1_000_000_000).to_i64
+          io << "ns"
+        end
+      end
+    end
+
+    class Sizes < Values(Int64)
+      KILOBYTE = 1024
+      MEGABYTE = 1024 * 1024
+      GIGABYTE = 1024 * 1024 * 1024
+
+      def humanize(io, value)
+        value = value.abs.to_u64
+
+        if value >= GIGABYTE
+          io << (value // GIGABYTE)
+          io << "GB"
+        elsif value >= MEGABYTE
+          io << (value // MEGABYTE)
+          io << "MB"
+        elsif value >= KILOBYTE
+          io << (value // KILOBYTE)
+          io << "KB"
+        else
+          io << value
+          io << 'B'
+        end
+      end
+    end
+
+    module Parser
+      def self.parse_event(line : String) : Event?
+        reader = Char::Reader.new(line)
+        section = parse_word
+        expect ' '
+        operation = parse_word
+        expect ' '
+        expect 't'
+        expect '='
+        time = parse_integer / 1_000_000_000 # nanoseconds
+        expect ' '
+        expect 'd'
+        expect '='
+        duration = parse_integer / 1_000_000_000 # nanoseconds
+        Event.new(section, operation, time, duration, line)
+      end
+
+      def self.parse_variable(line : String, name : String) : Int64?
+        if pos = line.index(name)
+          reader = Char::Reader.new(line, pos + name.bytesize)
+          expect '='
+          parse_integer
+        end
+      end
+
+      # Tries to parse known words, so we can return static strings instead of
+      # dynamically allocating the same string over an over, then falls back to
+      # allocate a string.
+      private macro parse_word
```

Comment:
I feel these many macros can be just defs. What's the reason for them being macros?

Reply:
This is mostly for early returns, since Parser could become a struct or char, and Char::Reader be an ivar... I'll try that. Dealing with return on each `parse_*` will be boring but acceptable.